R/req-perform-stream.R

```diff
@@ -52,6 +52,40 @@ req_perform_stream <- function(req, callback, timeout_sec = Inf, buffer_kb = 64)
   )
 }
 
+#' @param n number of lines to read from the stream
+#' @rdname req_perform_stream
+#' @export
+req_perform_stream_lines <- function(req, callback, timeout_sec = Inf, n = 1L) {
+  check_request(req)
+
+  handle <- req_handle(req)
+  callback <- as_function(callback)
+
+  stopifnot(is.numeric(timeout_sec), timeout_sec > 0)
+  stop_time <- Sys.time() + timeout_sec
+
+  stream <- curl::curl(req$url, handle = handle)
+  open(stream, "rbf")
+  withr::defer(close(stream))
+
+  continue <- TRUE
+  while(continue && isIncomplete(stream) && Sys.time() < stop_time) {
+    buf <- readLines(stream, n = n)
```

Comment:
Do you know what happens if there is data on the stream but not a newline? Does it block until a newline occurs?

Reply:
🤔 I'll check with a curl + webfakes but on a `rawConnection()` it returns with a warning. That's probably not ideal:

``` r
con <- rawConnection(charToRaw("greetings without a new line"))
readLines(con, 1)
#> Warning in readLines(con, 1): incomplete final line found on
#> 'charToRaw("greetings without a new line")'
#> [1] "greetings without a new line"
```

<sup>Created on 2024-02-09 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

Reply:
(I don't know how to simulate incomplete lines, but):

It looks like `readLines()` returns with less lines than requested when there is not enough in the stream. So I suppose this would need some sort of buffering if we wanted to make it blocking. 

``` r
library(magrittr)
suppressPackageStartupMessages(attach(asNamespace("httr2")))

req <- request(example_url()) %>% 
  req_url_path("/stream/2")
  
handle <- req_handle(req)
stream <- curl::curl(req$url, handle = handle)
open(stream, "rt")

readLines(stream, 5)
#> [1] "{\"url\":\"http://127.0.0.1:49833/stream/2\",\"args\":{},\"headers\":{\"Host\":\"127.0.0.1:49833\",\"User-Agent\":\"httr2/1.0.0.9000 r-curl/5.2.0 libcurl/7.84.0\",\"Accept\":\"*/*\",\"Accept-Encoding\":\"deflate, gzip\"},\"origin\":\"127.0.0.1\",\"id\":0}"
#> [2] "{\"url\":\"http://127.0.0.1:49833/stream/2\",\"args\":{},\"headers\":{\"Host\":\"127.0.0.1:49833\",\"User-Agent\":\"httr2/1.0.0.9000 r-curl/5.2.0 libcurl/7.84.0\",\"Accept\":\"*/*\",\"Accept-Encoding\":\"deflate, gzip\"},\"origin\":\"127.0.0.1\",\"id\":1}"
readLines(stream, 3)
#> character(0)
```

<sup>Created on 2024-02-09 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

Reply:
Perhaps this could be done "from the outside" with some sort of adaptor. Something like this: 

``` r
library(httr2)

stream_line <- function(fun) {
  buffer <- c()
  
  callback = function(x) {
    buffer <<- c(buffer, x)
    
    repeat {
      new_lines <- which(buffer == charToRaw("\n"))
      if (!length(new_lines)) break
      
      pos <- new_lines[1L]
      line <- rawToChar(head(buffer, pos - 1L))
      buffer <<- tail(buffer, -pos)
      
      if (!fun(line)) {
        return(FALSE)
      }
    }
    
    TRUE
  }
  
  environment()
}

fun <- function(x) {
  if (length(x)) {
    print(x)
  }
  invisible(TRUE)
}
adaptor <- stream_line(fun)
resp <- request(example_url()) |>
  req_url_path("/stream/10") |>
  req_perform_stream(adaptor$callback, buffer_kb = 1)
#> [1] "{\"url\":\"http://127.0.0.1:65493/stream/10\",\"args\":{},\"headers\":{\"Host\":\"127.0.0.1:65493\",\"User-Agent\":\"httr2/1.0.0.9000 r-curl/5.2.0 libcurl/7.84.0\",\"Accept\":\"*/*\",\"Accept-Encoding\":\"deflate, gzip\"},\"origin\":\"127.0.0.1\",\"id\":0}"
#> [1] "{\"url\":\"http://127.0.0.1:65493/stream/10\",\"args\":{},\"headers\":{\"Host\":\"127.0.0.1:65493\",\"User-Agent\":\"httr2/1.0.0.9000 r-curl/5.2.0 libcurl/7.84.0\",\"Accept\":\"*/*\",\"Accept-Encoding\":\"deflate, gzip\"},\"origin\":\"127.0.0.1\",\"id\":1}"
#> [1] "{\"url\":\"http://127.0.0.1:65493/stream/10\",\"args\":{},\"headers\":{\"Host\":\"127.0.0.1:65493\",\"User-Agent\":\"httr2/1.0.0.9000 r-curl/5.2.0 libcurl/7.84.0\",\"Accept\":\"*/*\",\"Accept-Encoding\":\"deflate, gzip\"},\"origin\":\"127.0.0.1\",\"id\":2}"
#> [1] "{\"url\":\"http://127.0.0.1:65493/stream/10\",\"args\":{},\"headers\":{\"Host\":\"127.0.0.1:65493\",\"User-Agent\":\"httr2/1.0.0.9000 r-curl/5.2.0 libcurl/7.84.0\",\"Accept\":\"*/*\",\"Accept-Encoding\":\"deflate, gzip\"},\"origin\":\"127.0.0.1\",\"id\":3}"
#> [1] "{\"url\":\"http://127.0.0.1:65493/stream/10\",\"args\":{},\"headers\":{\"Host\":\"127.0.0.1:65493\",\"User-Agent\":\"httr2/1.0.0.9000 r-curl/5.2.0 libcurl/7.84.0\",\"Accept\":\"*/*\",\"Accept-Encoding\":\"deflate, gzip\"},\"origin\":\"127.0.0.1\",\"id\":4}"
#> [1] "{\"url\":\"http://127.0.0.1:65493/stream/10\",\"args\":{},\"headers\":{\"Host\":\"127.0.0.1:65493\",\"User-Agent\":\"httr2/1.0.0.9000 r-curl/5.2.0 libcurl/7.84.0\",\"Accept\":\"*/*\",\"Accept-Encoding\":\"deflate, gzip\"},\"origin\":\"127.0.0.1\",\"id\":5}"
#> [1] "{\"url\":\"http://127.0.0.1:65493/stream/10\",\"args\":{},\"headers\":{\"Host\":\"127.0.0.1:65493\",\"User-Agent\":\"httr2/1.0.0.9000 r-curl/5.2.0 libcurl/7.84.0\",\"Accept\":\"*/*\",\"Accept-Encoding\":\"deflate, gzip\"},\"origin\":\"127.0.0.1\",\"id\":6}"
#> [1] "{\"url\":\"http://127.0.0.1:65493/stream/10\",\"args\":{},\"headers\":{\"Host\":\"127.0.0.1:65493\",\"User-Agent\":\"httr2/1.0.0.9000 r-curl/5.2.0 libcurl/7.84.0\",\"Accept\":\"*/*\",\"Accept-Encoding\":\"deflate, gzip\"},\"origin\":\"127.0.0.1\",\"id\":7}"
#> [1] "{\"url\":\"http://127.0.0.1:65493/stream/10\",\"args\":{},\"headers\":{\"Host\":\"127.0.0.1:65493\",\"User-Agent\":\"httr2/1.0.0.9000 r-curl/5.2.0 libcurl/7.84.0\",\"Accept\":\"*/*\",\"Accept-Encoding\":\"deflate, gzip\"},\"origin\":\"127.0.0.1\",\"id\":8}"
#> [1] "{\"url\":\"http://127.0.0.1:65493/stream/10\",\"args\":{},\"headers\":{\"Host\":\"127.0.0.1:65493\",\"User-Agent\":\"httr2/1.0.0.9000 r-curl/5.2.0 libcurl/7.84.0\",\"Accept\":\"*/*\",\"Accept-Encoding\":\"deflate, gzip\"},\"origin\":\"127.0.0.1\",\"id\":9}"
fun(adaptor$buffer)
```

<sup>Created on 2024-02-09 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>